### PR TITLE
Hyb 694 error modal

### PR DESCRIPTION
**Pull request by tony-dinh**
Error Modal Scaffold

JIRA: [**[HYB-694] As a scaffold user, I want a no internet / error screen**](https://mobify.atlassian.net/browse/HYB-694)
Linked PRs: **None**

## Changes
- Error modals appear when navigating without internet
- Error modals reveal back button when the current view can navigate
- Error modal content is customizable through `errorConfig.js`

## How to test-drive this PR
- Run app without internet: error modal should appear
- Leave internet off, tap retry: error modal should reappear
- Turn on internet, tap retry: modal should dismiss & tab/drawer items should be loaded
- Navigate a level deeper, turn off internet and navigate: back button should appear when possible to navigate back
- Open cart without internet: error modal should appear & back button should dismiss both cart & error modal

## TODOS:
- [x] Change works in both Android and iOS.
- [x] CHANGELOG.md has been updated.
- [x] Scaffold is working. If a new feature was added, it was added to the Scaffold app.
- [x] Run the generator to make sure it still functions correctly.
- [ ] +1 from an engineer on the Astro team.
- [x] Both tab layout and drawer layout are fully functional in ios. (Set `iosUsingTabLayout` in `baseConfig`)

### CODE REVIEW

File: app/app.js [hunk 1]
```diff
@@ -6,69 +6,100 @@ window.run = function() {
...
-                return Application.setMainViewPlugin(layout);
-            });
-
-            // Tab layout must be added as the mainViewPlugin before
-            // The first tab is selected or else the navigation does
-            // not complete correctly
-            return Promise.join(tabBarControllerPromise, layoutSetupPromise,
-            function(tabBarController) {
+                return tabBarController;
+            }).then(function(tabBarController) {
```
Comment (jeremywiebe) [inline app/app.js]: Do we need to return the `tabBarController` here and then use `.then()` right next? It seems like we could eliminate these two lines and just call `selectTab` right away.

File: app/scaffold-controllers/error-screen/errorController.js [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+    ErrorController.prototype.errorContent = function() {
+        if (!this.errorType) {
+            return null;
+        }
+        return ErrorConfig.errors[this.errorType];
+    };
+
+    // Remove the events once the modal is hidden or else some untriggered
+    // once events will still exist on the errorModal page and create
```
Comment (jeremywiebe) [inline app/scaffold-controllers/error-screen/errorController.js]: Nitpick... can we quote "once" ... I had to read this 3 times to understand we were talking about `once()` events. :)

File: app/scaffold-controllers/error-screen/errorController.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    // inconsitant behaviour in the navigation/web views
```
Comment (jeremywiebe) [inline app/scaffold-controllers/error-screen/errorController.js]: `s/inconsitant/inconsistant/`

File: app/scaffold-controllers/error-screen/errorController.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        this.viewPlugin.off('retry');
+        this.viewPlugin.off('back');
+    };
+
+    ErrorController.prototype._generateErrorCallback = function(errorType, params) {
+        var self = this;
+        var backHandler = params.backHandler;
+        var retryHandler = params.retryHandler;
+        var isActiveItem = params.isActiveItem;
+        return function(eventArgs) {
```
Comment (jeremywiebe) [inline app/scaffold-controllers/error-screen/errorController.js]: Purely a style comment, but I feel like an empty line before this return statement would make it visually easy to see we're returning a function (it wouldn't blend with the previous `var`'s as much).

File: app/scaffold-controllers/error-screen/errorController.js [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+            if (isActiveItem()) {
+                // Wait until the error page is loaded before showing
+                self.viewPlugin.on('error:loaded', function() {
```
Comment (jeremywiebe) [inline app/scaffold-controllers/error-screen/errorController.js]: Would it be worth making `error:loaded` a constant?

File: app/scaffold-controllers/tabBarController.js [hunk 6]
```diff
@@ -26,7 +26,13 @@ function(
         return tabBar;
     };
 
-    var initRegularTabs = function(tabBar, tabItems, cartEventHandler, counterBadgeController) {
+    var initRegularTabs = function(
```
Comment (jeremywiebe) [inline app/scaffold-controllers/tabBarController.js]: :clap:

File: app/scaffold-controllers/tabBarController.js [hunk 7]
```diff
@@ -89,5 +112,11 @@ function(
         this.tabBar.NavigationControllers[this.activeTabId].navigate(url);
     };
 
+    TabBarController.prototype.canGoBack = function() {
+    var activeTab = this.tabBar.NavigationControllers[this.activeTabId];
```
Comment (jeremywiebe) [inline app/scaffold-controllers/tabBarController.js]: Looks like indentation here went a bit wonky.

File: app/scaffold-controllers/error-screen/errorController.js (hunk 5 above)
Comment (tony-dinh) [inline app/scaffold-controllers/error-screen/errorController.js]: Hmm I can't think of benefits of making it a constant other than improving the readability of the code... In which case, I think that it fairly self-documented already, but if there are other benefits then we can definitely make it a constant!

Comment (jeremywiebe) [inline app/scaffold-controllers/error-screen/errorController.js]: I guess I usually lean towards a constant if I see a string literal in more than one place. :)

**Comment by jeremywiebe** [comment]
:+1:  LGTM!  :fireworks:

**Pull request merged**